sklearn/compose/_column_transformer.py

```diff
@@ -509,6 +517,16 @@ def transform(self, X):
         """
         check_is_fitted(self, 'transformers_')
 
+        # No column reordering allowed for named cols combined with remainder
+        if self._remainder[2] is not None and hasattr(self, '_df_columns'):
+            n_cols_fit = len(self._df_columns)
+            n_cols_transform = len(X.columns)
```

Comment:
This will raise an error if `X` is a numpy array. Check if `X` is a pandas array before heading into this path?

In future PR, we can use `_df_columns` to signal that `X` was a dataframe during `fit` and raise an error if during `transform`, `X` is not a dataframe.

Reply:
A good point, @thomasjpfan

Reply:
Good catch, thanks! Duck-typing check ok?
```python
if self._remainder[2] is not None and hasattr(self, '_df_columns') and hasattr(X, 'columns'):
```

Reply:
Yes, though ideally we should also be checking the number of columns when X is not a DataFrame. Not sure if that needs to be in this pull request

Reply:
Interpreting the last suggestion loosely, I looked into also checking column order for structured numpy arrays (np arrays with named columns) in an analogous way as now for DataFrames and I found that this would change the PR a lot. ColumnTransformer currently explicitly does not allow named columns for non-DataFrames and raises an error if one tries. I now think that this was not intended, but wanted to clearly state my (maybe misguided) interpretations. If this train of thought is relevant, we can continue this discussion in Andreas' RFC #14251.

When reading the suggestion literally, i.e. *only* checking the number of columns for non-DataFrames -- this condition is only in the code originally to avoid an exception when referencing X.columns. If this check is a goal of the PR itself, I can change the conditionals (and it would become a somewhat separate check with a separate error message as `_df_columns` only exists for DataFrames combined with column names). I can gladly do this if it's found useful.

For now, I'll only commit the suggested change, but that does not mean that I wouldn't do the other(s) at all. Just waiting for a yay or nay on those. :)

Reply:
PS to the commit: Note that when passing an np array to transform, an error will still be raised, not here, but in the correct location in the code (ColumnTransformer if named columns are used with non-DataFrames and raises a ValueError if not).

Reply:
We explicitly decided not to support struct arrays. The proposal here was to make sure we continue to allow fitting on DataFrame and transforming on numpy array, if that is currently supported.

The check for the number of columns in an array is because we provide that security in every other estimator, raising an error if the input has changed shape. Here it is also applicable but not currently done. It's not really necessary for this pull request

Reply:
Thanks @jnothman! I interpret your answer as "it's up to me." ;)

I added a general `self.n_features_` check in analogy to this one https://github.com/scikit-learn/scikit-learn/blob/7b8cbc875b862ebb81a9b3415bdee235cca99ca6/sklearn/ensemble/bagging.py#L682 but tweaked it to allow for added columns (as discussed [above](https://github.com/scikit-learn/scikit-learn/pull/14237#discussion_r300511521)). I expanded the tests for accepting same or more and rejecting fewer columns.

Referencing relevant PR #13603 as `ColumnTransformer` now also has `n_features_`

Reply:
> but tweaked it to allow for added columns (as discussed above)

Things get a little hairy here. I feel like for numpy arrays we should not be that lenient, since they are arrays not columnar frames...

But I can accept this for the version 0.20-21 fix.

> Referencing relevant PR #13603 as ColumnTransformer now also has n_features_

Similarly, since I'm trying to include this in a patch release, make this `_n_features` to keep the patch API compatible. This also avoids making more deprecation work in #13603 or subsequent to it.

Reply:
> > but tweaked it to allow for added columns (as discussed above)
> 
> Things get a little hairy here. I feel like for numpy arrays we should not be that lenient, since they are arrays not columnar frames...
> 
> But I can accept this for the version 0.20-21 fix.

Ok, if it is acceptable as it is for this fix, I'm leaving it unchanged (both arrays as well as DFs are checked leniently, accepting added columns, but not removed columns). That way I'm also confident that this PR doesn't break any relied-on behavior. Looking forward to other PRs like #14251 and #13603 for contributing to more consistent and clear checks.

> > Referencing relevant PR #13603 as ColumnTransformer now also has n_features_
> 
> Similarly, since I'm trying to include this in a patch release, make this `_n_features` to keep the patch API compatible. This also avoids making more deprecation work in #13603 or subsequent to it.

I see, thanks! I renamed it to `_n_features`, this is the only change in the latest commit.